include/interfaces/stokes_nc.h

```diff
@@ -0,0 +1,292 @@
+                                        const TrilinosWrappers::BlockSparseMatrix &,
+                                        LinearOperator<VEC> &,
+                                        LinearOperator<VEC> &) const;
+
+private:
+  double eta;
+
+  mutable shared_ptr<TrilinosWrappers::PreconditionAMG>    Amg_preconditioner;
+  mutable shared_ptr<TrilinosWrappers::PreconditionJacobi> Mp_preconditioner;
+  mutable shared_ptr<TrilinosWrappers::PreconditionJacobi> T_preconditioner;
+
+};
+
+template<int dim>
+StokesNC<dim>::StokesNC() :
+  NonConservativeInterface<dim,dim,dim+1,StokesNC<dim> >("StokesNC",
+                                                         "FESystem[FE_Q(2)^d-FE_Q(1)]",
+                                                         "u,u,p", "1,1; 1,0", "1,0; 0,1","0,0")
+{};
+
+
+template <int dim>
+template<typename Number>
+void StokesNC<dim>::preconditioner_residual(const typename DoFHandler<dim>::active_cell_iterator &cell,
+                                            Scratch &fe_cache,
+                                            CopyPreconditioner &data,
+                                            std::vector<Number> &local_residual) const
+{
+  Number alpha = this->alpha;
+  fe_cache.reinit(cell);
+
+  fe_cache.cache_local_solution_vector("solution",      *this->solution,      alpha);
+  fe_cache.cache_local_solution_vector("solution_dot",  *this->solution_dot,  alpha);
+
+  this->fix_solution_dot_derivative(fe_cache, alpha);
+
+  const FEValuesExtractors::Vector velocity(0);
+  const FEValuesExtractors::Scalar pressure(dim);
+
+  auto &us      = fe_cache.get_values(    "solution",     "u",      velocity,       alpha);
+  auto &grad_us = fe_cache.get_gradients( "solution",     "grad_u", velocity,       alpha);
+  auto &us_dot  = fe_cache.get_values(    "solution_dot", "u_dot",  velocity,       alpha);
+  auto &ps      = fe_cache.get_values(    "solution",     "p",      pressure,       alpha);
+
+  const unsigned int n_q_points = ps.size();
+
+  // Jacobian:
+  auto &JxW = fe_cache.get_JxW_values();
+
+  // Init residual to 0
+  for (unsigned int i=0; i<local_residual.size(); ++i)
+    local_residual[i] = 0;
+
+  auto &fev = fe_cache.get_current_fe_values();
+  for (unsigned int q=0; q<n_q_points; ++q)
+    {
+      for (unsigned int i=0; i<local_residual.size(); ++i)
+        {
+          // test functions:
+          auto v      = fev[velocity    ].value(i,q);
+          auto grad_v = fev[velocity    ].gradient(i,q);
+          auto m      = fev[pressure    ].value(i,q);
+
+          // variables:
+          const Tensor<1, dim, Number> &u       = us[q];
+          const Tensor<1, dim, Number> &u_dot   = us_dot[q];
+          const Tensor<2, dim, Number> &grad_u  = grad_us[q];
+          const Number                 &p       = ps[q];
+
+          // compute residual:
+          local_residual[i] +=
+            (eta * inner(grad_u,grad_v) +
+             (1./eta)*p*m)*JxW[q];
+
```

Comment:
voglio risolvere
-delta u = f(t)
ci sono cose che non mi tornano e voglio essere sicuro che vada tutto prima di complicare le cose.
per esempio, mi sembra che le forzanti non vengano applicate in questo caso. non ne sono sicuro, sto facendo un po’ di prove.

> On 02 Nov 2015, at 22:29, Alberto Sartori <notifications@github.com> wrote:
> 
> In include/interfaces/stokes_nc.h <https://github.com/mathLab/pi-DoMUS/pull/23#discussion_r43685213>:
> 
> > +          // test functions:
> > +          auto v      = fev[velocity    ].value(i,q);
> > +          auto grad_v = fev[velocity    ].gradient(i,q);
> > +          auto m      = fev[pressure    ].value(i,q);
> > +
> > +          // variables:
> > +          const Tensor<1, dim, Number> &u       = us[q];
> > +          const Tensor<1, dim, Number> &u_dot   = us_dot[q];
> > +          const Tensor<2, dim, Number> &grad_u  = grad_us[q];
> > +          const Number                 &p       = ps[q];
> > +
> > +          // compute residual:
> > +          local_residual[i] +=
> > +            (eta * inner(grad_u,grad_v) +
> > +             (1./eta)*p*m)*JxW[q];
> > +
> non ti seguo. vuoi fare un problema quasi-statico?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/mathLab/pi-DoMUS/pull/23/files#r43685213>.
>